// ngx-particle/esm2015/lib/particles.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input } from '@angular/core';
let ParticlesComponent = class ParticlesComponent {
    constructor() {
        this.width = 100;
        this.height = 100;
        this.style = {};
    }
};
__decorate([
    Input(),
    __metadata("design:type", Number)
], ParticlesComponent.prototype, "width", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], ParticlesComponent.prototype, "height", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], ParticlesComponent.prototype, "params", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], ParticlesComponent.prototype, "style", void 0);
ParticlesComponent = __decorate([
    Component({
        selector: 'particles',
        template: `
        <div [ngStyle]="style" class="particles-container">
            <canvas d-particles [params]="params" [style.width.%]="width" [style.height.%]="height"></canvas>
        </div>
    `
    }),
    __metadata("design:paramtypes", [])
], ParticlesComponent);
export { ParticlesComponent };
