// ngx-particle/esm5/lib/particles.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input } from '@angular/core';
var ParticlesComponent = /** @class */ (function () {
    function ParticlesComponent() {
        this.width = 100;
        this.height = 100;
        this.style = {};
    }
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], ParticlesComponent.prototype, "width", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], ParticlesComponent.prototype, "height", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], ParticlesComponent.prototype, "params", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], ParticlesComponent.prototype, "style", void 0);
    ParticlesComponent = __decorate([
        Component({
            selector: 'particles',
            template: "\n        <div [ngStyle]=\"style\" class=\"particles-container\">\n            <canvas d-particles [params]=\"params\" [style.width.%]=\"width\" [style.height.%]=\"height\"></canvas>\n        </div>\n    "
        }),
        __metadata("design:paramtypes", [])
    ], ParticlesComponent);
    return ParticlesComponent;
}());
export { ParticlesComponent };
